Sum type with struct and function types

### Describe the bug

Recently discovered that Sum types don't allow Interfaces. Perhaps they don't allow function types either?

```
struct None {}
type Myfn = fn (int) int
type Myfnfact = fn () Myfn
type Maybefnfact = Myfnfact | None

//Myfn
fn abc(i int) int {
    return i
}

fn run(mmff Maybefnfact) {
    match mmff {
        Myfnfact {println('yes fn')}
        None {println('None fn')}
    }
}

//create Myfn
myfnfact := fn () Myfn {
    return abc
}

println(myfnfact()(1))
run(None{})	// works 
run(myfnfact) // no work
```

### Expected Behavior

Expected to see output from run() function.

### Current Behavior

```
/tmp/v_60000/code.16139365580916353913.tmp.c:2072: error: ',' expected (got "x")
builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

Exited with error status 1
```

### Reproduction Steps

Run the above code in playground.

### Possible Solution

Expand Sum types to include function types (and interfaces)...if that's the issue. Or give compiler message about the limitation.

### Additional Information/Context

_No response_

### V version

playground

### Environment details (OS name and version, etc.)

playground

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2285,6 +2285,17 @@ fn (mut g Gen) write_sumtype_casting_fn(fun SumtypeCastingFn) {
 			got_name := 'fn ${g.table.fn_type_source_signature(got_sym.info.func)}'
 			got_cname = 'anon_fn_${g.table.fn_type_signature(got_sym.info.func)}'
 			type_idx = g.table.type_idxs[got_name].str()
+			exp_info := exp_sym.info as ast.SumType
+			for variant in exp_info.variants {
+				variant_sym := g.table.sym(variant)
+				if variant_sym.info is ast.FnType {
+					if g.table.fn_type_source_signature(variant_sym.info.func) == g.table.fn_type_source_signature(got_sym.info.func) {
+						got_cname = variant_sym.cname
+						type_idx = variant.idx().str()
+						break
+					}
+				}
+			}
 			sb.writeln('static inline ${exp_cname} ${fun.fn_name}(${got_cname} x) {')
 			sb.writeln('\t${got_cname} ptr = x;')
 			is_anon_fn = true

--- a/vlib/v/tests/sumtype_with_alias_fntype_test.v
+++ b/vlib/v/tests/sumtype_with_alias_fntype_test.v
@@ -1,0 +1,37 @@
+struct None {}
+
+type Myfn = fn (int) int
+
+type Myfnfact = fn () Myfn
+
+type Maybefnfact = Myfnfact | None
+
+// Myfn
+fn abc(i int) int {
+	return i
+}
+
+fn run(mmff Maybefnfact) string {
+	match mmff {
+		Myfnfact { return 'yes fn' }
+		None { return 'None fn' }
+	}
+}
+
+fn test_sumtype_with_alias_fntype() {
+	// create Myfn
+	myfnfact := fn () Myfn {
+		return abc
+	}
+	r1 := main.myfnfact()(1)
+	println(r1)
+	assert r1 == 1
+
+	r2 := run(None{})
+	println(r2)
+	assert r2 == 'None fn'
+
+	r3 := run(myfnfact)
+	println(r3)
+	assert r3 == 'yes fn'
+}

